File: play-2.5.x/K-Modules/kmodules/K-Base/app/K/Service/PlanTasks/PlanTaskService.java
package K.Service.PlanTasks;

import K.Common.Helper;
import K.DataDict.TaskStatus;
import K.Ebean.DB;
import K.Service.ServiceBase;
import com.avaje.ebean.TxRunnable;
import jodd.datetime.JDateTime;
import jodd.exception.ExceptionUtil;
import models.K.BgTask.PlanTask;
import play.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class PlanTaskService extends ServiceBase {

    public static final String ServiceName = "PlanTaskService";

    private int task_loader_wait_time = 5;
    private boolean stop_now;

    private ScheduledExecutorService seq_worker;
    private ScheduledExecutorService parallel_worker;

    private Thread seq_task_loader;
    private Thread parallel_task_loader;

    private Object task_notifier;

    private int parallel_worker_thread_count = 4;


    public PlanTaskService() {
        super(ServiceName);
        task_notifier = new Object();
        stop_now = false;
        PlanTask.ResetTaskStatus();
    }

    private void Init() {
        parallel_worker = Executors.newScheduledThreadPool(parallel_worker_thread_count);

        seq_worker = Executors.newSingleThreadScheduledExecutor();
    }

    @Override
    public void Start() {
        if (Running()) {
            return;
        }
        Init();
        stop_now = false;

        start_task_loader(true);
        start_task_loader(false);

        setRunning(true);
        Logger.debug("Plan Task Service Started......");
    }

    @Override
    public boolean Stop() {
        if (!Running()) {
            return true;
        }

        stop_now = true;
        Logger.debug("==> Try to stop plan task loader...");
        NotifyNewTask();

        try {
            parallel_task_loader.join(120000);
            seq_task_loader.join(120000);

            Logger.debug("==> Try to stop plan task worker...");
            parallel_worker.shutdown();
            seq_worker.shutdown();

            parallel_worker.awaitTermination(120, TimeUnit.SECONDS);
            seq_worker.awaitTermination(120, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            // todo : 增加错误日志
            Logger.warn(Helper.StackTraceOfEx(e));
            return false;
        }

        setRunning(false);
        Logger.debug("Plan Task Service Stopped......");
        return true;
    }

    private void process_task(final PlanTask task) {
        try {
            if (task != null) {
                final Runnable run_entity = (Runnable) Helper.FromJsonString(task.json_data,
                        Class.forName(task.class_name));

                if (run_entity != null) {
                    try {
                        DB.RunInTransaction(new TxRunnable() {
                            @Override
                            public void run() {
                                run_entity.run();
                                DB.Default().delete(task);
                            }
                        });
                    } catch (Exception ex) {
                        String ex_msg = ExceptionUtil.exceptionChainToString(ex);
                        Logger.error("Run Task: {} Failed.\nException: {}",
                                Helper.ToJsonStringPretty(task),
                                ex_msg);
                        DB.RunInTransaction(new TxRunnable() {
                            @Override
                            public void run() {
                                task.task_status = TaskStatus.Exception.code;
                                task.remarks = ex_msg;
                                DB.Default().save(task);
                            }
                        });

                    }
                }
            }
        } catch (Exception ex) {
            Logger.error(ExceptionUtil.exceptionChainToString(ex));
        }
    }

    private void start_task_loader(boolean require_seq) {
        final List<PlanTask> loaded_tasks = new ArrayList<>();

        final ScheduledExecutorService worker;
        if (require_seq) {
            worker = this.seq_worker;
        } else {
            worker = this.parallel_worker;
        }

        Thread task_loader = new Thread(new Runnable() {
            @Override
            public void run() {
                while (true) {
                    if (stop_now) return;

                    try {
                        // Load tasks to parallel_worker
                        DB.RunInTransaction(new TxRunnable() {
                            @Override
                            public void run() {
                                JDateTime end_time = new JDateTime();
                                end_time.addSecond(task_loader_wait_time + 1);

                                List<PlanTask> taskList = PlanTask.find.where()
                                        .eq("require_seq", require_seq)
                                        .eq("task_status", TaskStatus.WaitingInDB.code)
                                        .le("plan_run_time", end_time.convertToDate())
                                        .orderBy("id")
                                        .findList();

                                for (PlanTask task : taskList) {
                                    task.task_status = TaskStatus.WaitingInQueue.code;
                                    DB.Default().save(task);
                                }

                                loaded_tasks.clear();
                                loaded_tasks.addAll(taskList);
                            }
                        });

                        for (PlanTask task : loaded_tasks) {
                            SchedulePlanTask(task, worker);
                        }

                        if (loaded_tasks.size() == 0) {
                            try {
                                synchronized (task_notifier) {
                                    task_notifier.wait(task_loader_wait_time * 1000);
                                }
                            } catch (Exception ex) {
                                // do nothing
                            }
                        }

                    } catch (Exception e) {
                        Logger.error(Helper.StackTraceOfEx(e));
                    }
                }
            }
        });
        task_loader.start();

        if (require_seq) {
            this.seq_task_loader = task_loader;
        } else {
            this.parallel_task_loader = task_loader;
        }
    }

    private void SchedulePlanTask(final PlanTask task, final ScheduledExecutorService woker) {
        JDateTime now = new JDateTime();
        JDateTime plan_run_time = new JDateTime(task.plan_run_time);
        long interval = plan_run_time.getTimeInMillis() - now.getTimeInMillis();
        long delay = interval > 0 ? interval : 0;

        woker.schedule(new Runnable() {
                           @Override
                           public void run() {
                               try {
                                   process_task(task);
                               } catch (Exception ex) {
                                   Logger.error(ExceptionUtil.exceptionChainToString(ex));
                               }
                           }
                       },
                delay,
                TimeUnit.MILLISECONDS);
    }

    public void NotifyNewTask() {
        synchronized (task_notifier) {
            task_notifier.notifyAll();
        }
    }
}

File: play25kotlin/subProjects/ToolBox/app/k/task/PlanTaskService.kt
package k.task

import jodd.datetime.JDateTime
import jodd.exception.ExceptionUtil
import k.common.AnsiColor
import k.common.Helper
import k.common.Hub
import k.ebean.DB
import models.task.PlanTask
import models.task.TaskStatus
import play.Logger
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * Created by kk on 16/9/19.
 */
object PlanTaskService {

    private val seqWorker: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private val parallelWorker: ScheduledExecutorService = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors())

    private var seqTaskLoader: Thread? = null
    private var parallerTaskLoader: Thread? = null
    private val taskNotifier = Object()
    private val taskLoaderWaitTime = 5
    private var stopNow: Boolean = true

    var isRunning: Boolean = false
        private set

    fun Start() {
        if (isRunning) return

        stopNow = false

        seqTaskLoader = BuildTaskLoader(true, seqWorker)
        parallerTaskLoader = BuildTaskLoader(false, parallelWorker)

        seqTaskLoader!!.start()
        parallerTaskLoader!!.start()

        isRunning = true
        Helper.DLog(AnsiColor.GREEN, "Plan Task Service Started......")
    }

    fun Stop() {
        if (!isRunning) return
        stopNow = true
        try {
            Helper.DLog(AnsiColor.GREEN, "Try to stop plan task loader...")
            NotifyNewTask()
            seqTaskLoader!!.join(120000)
            parallerTaskLoader!!.join(120000)

            Helper.DLog(AnsiColor.GREEN, "Try to stop plan task worker...")
            seqWorker.shutdown()
            parallelWorker.shutdown()

            seqWorker.awaitTermination(120, TimeUnit.SECONDS)
            parallelWorker.awaitTermination(120, TimeUnit.SECONDS)

            Helper.DLog(AnsiColor.GREEN, "Plan Task Service Stopped......")
        } catch (ex: Exception) {
            Logger.error(ExceptionUtil.exceptionChainToString(ex))
        } finally {
            isRunning = false
        }
    }

    private fun BuildTaskLoader(requireSeq: Boolean, worker: ScheduledExecutorService): Thread {
        return Thread(Runnable {
            val loadedTasks = mutableListOf<PlanTask>()
            while (true) {
                if (stopNow) return@Runnable

                try {
                    DB.RunInTransaction {
                        val endTime = JDateTime().addSecond(taskLoaderWaitTime + 1)
                        val tasks = PlanTask.where()
                                .eq("require_seq", requireSeq)
                                .eq("task_status", TaskStatus.WaitingInDB.code)
                                .le("plan_run_time", endTime)
                                .findList()

                        tasks.forEach {
                            it.task_status = TaskStatus.WaitingInQueue.code
                        }
                        DB.Default().saveAll(tasks)

                        loadedTasks.clear()
                        loadedTasks.addAll(tasks)
                    }   // 事务截至点

                    if (loadedTasks.size > 0) {
                        loadedTasks.forEach {
                            SchedulePlanTask(it, worker)
                        }
                    } else {
                        // 在 endTime 之前没有需要执行的 task, 尝试等待新任务, 释放 cpu
                        try {
                            synchronized(taskNotifier, {
                                taskNotifier.wait(taskLoaderWaitTime * 1000L)
                            })
                        } catch (ex: Exception) {
                            // do nothing
                        }
                    }

                } catch (ex: Exception) {
                    loadedTasks.clear()
                    Logger.error(ExceptionUtil.exceptionChainToString(ex))
                }
            }
        })
    }

    private fun SchedulePlanTask(task: PlanTask, worker: ScheduledExecutorService) {
        val now = JDateTime()
        if (task.plan_run_time == null) task.plan_run_time = now
        val interval = task.plan_run_time!!.timeInMillis - now.timeInMillis
        val delay = if (interval > 0) interval else 0

        worker.schedule({
            try {
                process_task(task)
            } catch (ex: Exception) {
                Logger.error(ExceptionUtil.exceptionChainToString(ex))
            }
        },
                delay,
                TimeUnit.MILLISECONDS)
    }

    private fun DeserializeJsonData(task: PlanTask): Runnable? {
        try {
            return Helper.FromJsonString(task.json_data!!, Class.forName(task.class_name)) as Runnable
        } catch (ex: Exception) {
            return null
        }
    }

    private fun process_task(task: PlanTask) {
        try {
            val runObj = DeserializeJsonData(task)
            if (runObj != null) {
                try {
                    DB.RunInTransaction({
                        runObj.run()    // 执行任务
                        task.refresh()
                        task.delete()   // 任务执行成功后, 从数据库里删除记录
                    })
                } catch (ex: Exception) {
                    // 任务执行发生错误, 标记任务状态, 记录
                    DB.RunInTransaction {
                        task.refresh()
                        task.task_status = TaskStatus.Error.code
                        task.remarks = ExceptionUtil.exceptionChainToString(ex)
                        task.save()
                    }
                }
            } else {
                DB.RunInTransaction {
                    task.refresh()
                    task.task_status = TaskStatus.Error.code
                    task.remarks = "反序列化任务失败"
                    task.save()
                }
            }
        } catch (ex: Exception) {
            Logger.error(ExceptionUtil.exceptionChainToString(ex))
        }
    }

    fun NotifyNewTask() {
        synchronized(taskNotifier, {
            taskNotifier.notifyAll()
        })
    }

    fun Enabled(): Boolean {
        return Hub.configuration().getBoolean("k.planTaskService", false)
    }
}